compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2494,11 +2494,15 @@ impl<'a> Parser<'a> {
         // Parse the arguments, starting out with `self` being allowed...
         let (mut params, _) = self.parse_paren_comma_seq(|p| {
             p.recover_diff_marker();
+            let starts_with_paren = p.check_noexpect(&token::OpenDelim(Delimiter::Parenthesis));
             let param = p.parse_param_general(req_name, first_param).or_else(|mut e| {
                 e.emit();
                 let lo = p.prev_token.span;
                 // Skip every token until next possible arg or end.
                 p.eat_to_tokens(&[&token::Comma, &token::CloseDelim(Delimiter::Parenthesis)]);
```

Comment:
Ok, I think this logic is just incorrect.
We don't need to eat tokens until the first `,` or `)`.
What we really need here is to eat tokens until all delimiters (starting from after `recover_diff_marker` above) are balanced, and *then* eat tokens until the first `,` or `)`.

Reply:
Not sure how to do that exactly, but the easiest way to do that is probably to use `create_snapshot_for_diagnostic` after `recover_diff_marker`, and then roll back to that position in case of an error and eat token trees until `,` or `EOF`.

Reply:
It works, but why? Could you tell me the difference? Thanks!

Reply:
I think I may have figured it out. For example, for `fn foo((i32, i32) (a, b))`, the position may reach `)` after `b` by mistake, right?

Reply:
Yes, something like that.
`parse_param_general` stops at some unexpected token inside the parameter, so we first need to exit the parameter and only then start searching for the next parameter (`,`) or the end of parameter list (`)`).
If we start searching without exiting first we can find some arbitrary incorrect `,` or `)` inside the parameter.